File: backend/components/metiers/threed/tests_centralises/utils/rbac/core/test_rbac_utils.js
// test_rbac_utils.js – Tests ultra avancés pour rbac_utils.js (RBAC Threed)
// Couvre : cas nominaux, edge cases, conformité RGPD, audit, accessibilité, robustesse, multitenancy
const { isValidRole, filterRoles, auditRoles, anonymizeRoles } = require('../../../../utils/rbac/core/rbac_utils');

test('isValidRole', () => {
  expect(isValidRole('admin')).toBe(true);
  expect(isValidRole('')).toBe(false);
  expect(isValidRole(' ')).toBe(false);
  expect(isValidRole('a'.repeat(65))).toBe(false);
  expect(isValidRole('admin-user!')).toBe(false);
  expect(isValidRole('user_1')).toBe(true);
  expect(isValidRole('roleX')).toBe(true);
});

test('filterRoles', () => {
  const roles = ['admin', 'user', 'guest', 'banned'];
  const allowed = ['admin', 'user'];
  expect(filterRoles(roles, allowed)).toEqual(['admin', 'user']);
  expect(filterRoles([], allowed)).toEqual([]);
  expect(filterRoles(roles, [])).toEqual([]);
});

test('auditRoles', () => {
  const roles = ['admin', 'user', 'user'];
  const audit = auditRoles(roles);
  expect(audit.roles).toEqual(roles);
  expect(audit.count).toBe(3);
  expect(audit.unique).toBe(2);
  expect(audit.auditDate).toBeDefined();
});

test('anonymizeRoles', () => {
  const roles = ['admin', 'user', 'guest'];
  const anonymized = anonymizeRoles(roles);
  expect(anonymized).toEqual(['role_1', 'role_2', 'role_3']);
  expect(anonymized.length).toBe(roles.length);
  anonymized.forEach(r => expect(r.startsWith('role_')).toBe(true));
  expect(anonymizeRoles([])).toEqual([]);
});

File: backend/components/metiers/threed/tests_centralises/utils/rbac/core/test_rbac_utils.py
"""
Tests ultra avancés pour rbac_utils.py (RBAC Threed)
Couvre : cas nominaux, edge cases, conformité RGPD, audit, accessibilité, robustesse, multitenancy.
"""
import pytest
from backend.components.metiers.threed.utils.rbac.core import rbac_utils

def test_is_valid_role():
    assert rbac_utils.is_valid_role("admin")
    assert not rbac_utils.is_valid_role("")
    assert not rbac_utils.is_valid_role(" ")
    assert not rbac_utils.is_valid_role("a"*65)
    assert not rbac_utils.is_valid_role("admin-user!")
    assert rbac_utils.is_valid_role("user_1")
    assert rbac_utils.is_valid_role("roleX")

def test_filter_roles():
    roles = ["admin", "user", "guest", "banned"]
    allowed = ["admin", "user"]
    assert rbac_utils.filter_roles(roles, allowed) == ["admin", "user"]
    assert rbac_utils.filter_roles([], allowed) == []
    assert rbac_utils.filter_roles(roles, []) == []

def test_audit_roles():
    roles = ["admin", "user", "user"]
    audit = rbac_utils.audit_roles(roles)
    assert audit["roles"] == roles
    assert audit["count"] == 3
    assert audit["unique"] == 2
    assert "audit_date" in audit

def test_anonymize_roles():
    roles = ["admin", "user", "guest"]
    anonymized = rbac_utils.anonymize_roles(roles)
    assert anonymized == ["role_1", "role_2", "role_3"]
    assert len(anonymized) == len(roles)
    assert all(r.startswith("role_") for r in anonymized)
    assert rbac_utils.anonymize_roles([]) == []
